### PR TITLE
Fully refetch working beatmap when entering editor

### LINKED ISSUE #21794: Editor can falsely warn of a difficulty with the same name while swapping names
### Discussed in https://github.com/ppy/osu/discussions/21785

<div type='discussions-op-text'>

<sup>Originally posted by **Gabixel** December 25, 2022</sup>
Not sure how to always reproduce this but most of the time, when swapping names to 2 difficulties, the editor gives a false error saying that one of the two difficulties already has the new name.
I am not convinced of this, but it seems like it could also depend on how soon this swap is made

Here are the steps I made in the video:
- you can see 2 difficulties:
  - 'easy' (blue)
  - 'normal' (cyan)
- renamed 'easy' to 'hard'
- renamed 'normal' to 'easy'
- tried to rename 'hard' to 'normal'.
  the editor says that the map already has a difficulty with the name 'normal'

Note: I made one of the two difficulties using the 'exact copy' option, and then I added more objects to increase its star rating.

https://user-images.githubusercontent.com/43073074/209465564-141de3e6-01f5-4c4b-9ba6-0846386a6834.mp4</div>

### PR DESCRIPTION
**Pull request by bdach**
Closes https://github.com/ppy/osu/issues/21794.

I'm not actually super sure as to what the exact mode of failure is here, but it's 99% to do with working beatmap cache invalidation. Likely this can be even considered as another case of https://github.com/ppy/osu/issues/21357, but because this is a one-liner "fix," I'm PRing it anyways. I spent about 1.5 hours too much on this already anyways compared to how long this _should_ have taken to debug.

The issue is confusing to understand when working with the swap scenario given in the issue, but it's a little easier to understand when performing the following:

1. Have a beatmap set with 2 difficulties. Let's call them "A" and "B".
2. From song select, **without ever exiting to main menu**, edit "A". Change the difficulty name to "AA". Save and exit back to song select; **do not exit out to main menu**.
3. From song select, edit "B". Change the difficulty name to "BB". Save and exit back to song select.
4. The difficulty names will be "A" and "BB".

Basically what I *think* is causing this, is the fact that even though editor invalidates the working beatmap by refetching it afresh on exit, song select is blissfully unaware of this, and continues working with its own `BeatmapInfo` instances which have backlinks to `BeatmapSetInfo`.

When editing the first of the two difficulties and then the second, the editing of the first one only invalidates the first one rather than the entire set, and the second difficulty continues to have a stale reference to the first one via the beatmap set with the old name, and as such ends up overwriting the changes from the first save when passed into the editor and modified again.

No tests because it's annoying to write one. I'll try on request.

### CODE REVIEW

**Comment by peppy** [comment]
Not against making this change, but can you confirm that it doesn't affect other usages of song select, ie. entering player state via
 https://github.com/ppy/osu/blob/5c049feca1e855daca7499b2e6bfb6482c99fe8a/osu.Game/Screens/Select/SongSelect.cs#L622

**Comment by bdach** [comment]
> can you confirm that it doesn't affect other usages of song select, ie. entering player state

Affect in what way? Flushing working beatmap cache and refetching anew can only ever be _more_ correct than not, can't it? Am I missing something?

This is also specifically places in the method that precedes entering editor, too.

**Comment by peppy** [comment]
Never mind, on testing with your repro steps I better understand the fail case.

**Review by peppy — APPROVED** [review]
